File: backend/app/schemas/trip_window.py
from datetime import date as date_type
from datetime import datetime

from pydantic import BaseModel, Field, model_validator

from app.schemas.location import LocationType
from app.schemas.preferences import PreferenceProfile
from app.schemas.scoring import ColorProbabilities
from app.schemas.session import SunEvent

# Open-Meteo's forecast horizon — bounds how far out a trip window can run.
MAX_TRIP_WINDOW_DAYS = 16


class TripWindowRequest(BaseModel):
    lat: float
    lon: float
    event: SunEvent
    start_date: date_type
    end_date: date_type
    radius_km: float = Field(10.0, gt=0, le=100)
    place_types: list[LocationType]
    preferences: PreferenceProfile = PreferenceProfile()
    tz_name: str = "UTC"

    @model_validator(mode="after")
    def _validate_date_range(self) -> "TripWindowRequest":
        if self.end_date < self.start_date:
            raise ValueError("end_date must not be before start_date")
        if (self.end_date - self.start_date).days >= MAX_TRIP_WINDOW_DAYS:
            raise ValueError(f"trip window cannot exceed {MAX_TRIP_WINDOW_DAYS} days")
        return self


class TripWindowResult(BaseModel):
    location_id: str
    name: str
    type: LocationType
    distance_km: float
    date: date_type

    event_time: datetime
    recommended_arrival_offset_minutes: int
    best_viewing_window_start: datetime
    best_viewing_window_end: datetime

    visibility_likelihood: float
    cloud_cover_summary: str
    color_probabilities: ColorProbabilities
    rain_or_unsafe_alert: str | None
    preference_match_score: float
    explanation: str | None = None


class TripWindowResponse(BaseModel):
    best: TripWindowResult | None
    candidates_considered: int
    days_considered: int

File: client/lib/models/trip_window.dart
import 'location_type.dart';
import 'preference_profile.dart';
import 'session_request.dart';
import 'session_response.dart';

class TripWindowRequest {
  final double lat;
  final double lon;
  final SunEvent event;
  final DateTime startDate;
  final DateTime endDate;
  final double radiusKm;
  final List<LocationType> placeTypes;
  final PreferenceProfile preferences;
  final String tzName;

  const TripWindowRequest({
    required this.lat,
    required this.lon,
    required this.event,
    required this.startDate,
    required this.endDate,
    required this.radiusKm,
    required this.placeTypes,
    required this.preferences,
    this.tzName = 'UTC',
  });

  Map<String, dynamic> toJson() => {
        'lat': lat,
        'lon': lon,
        'event': event.apiValue,
        'start_date': _formatDate(startDate),
        'end_date': _formatDate(endDate),
        'radius_km': radiusKm,
        'place_types': placeTypes.map((t) => t.apiValue).toList(),
        'preferences': preferences.toJson(),
        'tz_name': tzName,
      };

  static String _formatDate(DateTime date) {
    final month = date.month.toString().padLeft(2, '0');
    final day = date.day.toString().padLeft(2, '0');
    return '${date.year}-$month-$day';
  }
}

class TripWindowResult {
  final String locationId;
  final String name;
  final LocationType type;
  final double distanceKm;
  final String date; // "YYYY-MM-DD", as returned by the backend

  final String eventTime;
  final int recommendedArrivalOffsetMinutes;
  final String bestViewingWindowStart;
  final String bestViewingWindowEnd;

  final double visibilityLikelihood;
  final String cloudCoverSummary;
  final ColorProbabilities colorProbabilities;
  final String? rainOrUnsafeAlert;
  final double preferenceMatchScore;
  final String? explanation;

  const TripWindowResult({
    required this.locationId,
    required this.name,
    required this.type,
    required this.distanceKm,
    required this.date,
    required this.eventTime,
    required this.recommendedArrivalOffsetMinutes,
    required this.bestViewingWindowStart,
    required this.bestViewingWindowEnd,
    required this.visibilityLikelihood,
    required this.cloudCoverSummary,
    required this.colorProbabilities,
    required this.rainOrUnsafeAlert,
    required this.preferenceMatchScore,
    required this.explanation,
  });

  factory TripWindowResult.fromJson(Map<String, dynamic> json) => TripWindowResult(
        locationId: json['location_id'] as String,
        name: json['name'] as String,
        type: LocationType.fromApiValue(json['type'] as String),
        distanceKm: (json['distance_km'] as num).toDouble(),
        date: json['date'] as String,
        eventTime: json['event_time'] as String,
        recommendedArrivalOffsetMinutes: json['recommended_arrival_offset_minutes'] as int,
        bestViewingWindowStart: json['best_viewing_window_start'] as String,
        bestViewingWindowEnd: json['best_viewing_window_end'] as String,
        visibilityLikelihood: (json['visibility_likelihood'] as num).toDouble(),
        cloudCoverSummary: json['cloud_cover_summary'] as String,
        colorProbabilities:
            ColorProbabilities.fromJson(json['color_probabilities'] as Map<String, dynamic>),
        rainOrUnsafeAlert: json['rain_or_unsafe_alert'] as String?,
        preferenceMatchScore: (json['preference_match_score'] as num).toDouble(),
        explanation: json['explanation'] as String?,
      );
}

class TripWindowResponse {
  final TripWindowResult? best;
  final int candidatesConsidered;
  final int daysConsidered;

  const TripWindowResponse({
    required this.best,
    required this.candidatesConsidered,
    required this.daysConsidered,
  });

  factory TripWindowResponse.fromJson(Map<String, dynamic> json) => TripWindowResponse(
        best: json['best'] == null
            ? null
            : TripWindowResult.fromJson(json['best'] as Map<String, dynamic>),
        candidatesConsidered: json['candidates_considered'] as int,
        daysConsidered: json['days_considered'] as int,
      );
}